sdb/commands/linux/internal/slub_helpers.py

```diff
@@ -122,3 +113,160 @@ def util(cache: drgn.Object) -> int:
     if total_mem == 0:
         return 0
     return int((active_memory(cache) / total_mem) * 100)
+
+
+def cache_red_left_padding(cache: drgn.Object) -> int:
+    SLAB_RED_ZONE_FLAG = 0x00000400
+    padding = 0
+    if cache.flags & SLAB_RED_ZONE_FLAG:
+        padding += int(cache.red_left_pad)
+    return padding
+
+
+def cache_get_free_pointer(cache: drgn.Object, p: drgn.Object) -> drgn.Object:
+    """
+    Get the next pointer in the freelist. Note, that this
+    function assumes that CONFIG_SLAB_FREELIST_HARDENED
+    is set in the target
+    """
+    assert cache.type_.type_name() == 'struct kmem_cache *'
+    assert p.type_.type_name() == 'void *'
+    hardened_ptr = p + cache.offset.value_()
+
+    #
+    # We basically do what `freelist_dereference()` and
+    # `freelist_ptr()` do in the kernel source:
+    #
+    # ptr <- (void *)*(unsigned long *)(hardened_ptr)
+    #
+    intermediate_ulong = drgn.Object(sdb.get_prog(),
+                                     type='unsigned long',
+                                     address=hardened_ptr.value_())
+    ptr = drgn.Object(sdb.get_prog(),
+                      type='void *',
+                      value=intermediate_ulong.value_())
+
+    #
+    # ptr_addr <- (unsigned long)hardened_ptr
+    #
+    ptr_addr = drgn.Object(sdb.get_prog(),
+                           type='unsigned long',
+                           value=hardened_ptr.value_())
+
+    #
+    # return (void *)((unsigned long)ptr ^ cache->random ^ ptr_addr)
+    #
+    ptr_as_ulong = drgn.Object(sdb.get_prog(),
+                               type='unsigned long',
+                               value=ptr.value_())
+    clean_ptr_val = ptr_as_ulong.value_()
+    clean_ptr_val ^= cache.random.value_()
+    clean_ptr_val ^= ptr_addr.value_()
+    return drgn.Object(sdb.get_prog(), type='void *', value=clean_ptr_val)
+
+
+def cache_get_free_set(cache: drgn.Object,
+                       page: drgn.Object) -> Iterable[drgn.Object]:
```

Comment:
It seems like the convention for how to name a function that gets all X given a Y is `for_each_X_in_Y`, so I think this should be `for_each_freeobj_in_slab`.

Reply:
done and good idea - we can start standardizing this now. I'll probably do this and all the yielding refactoring in spl_kmem_cache walker after this lands.